### ISSUE TITLE
Apple GPU via jax-mps: reversed-array times broadcast scalar is silently zeroed under MLX compile (jax-mps#232); polar_dust.py's reversed trapezoid trips it, three more trapezoid sites are candidates (#1206)

### ISSUE THREAD

**Issue by suchethac** [suchethac/tengri#2295]
**Child of #1206; upstream defect https://github.com/tillahoffmann/jax-mps/issues/232.**

On Apple GPU via jax-mps 0.10.10 with MLX kernel compilation on (the default), a reversed array combined with a broadcast scalar is silently wrong: `y[::-1] * 2.0` keeps element 0 and zeros the rest (`jnp.flip`, `lax.rev`, any negative step; float32 and int32; jitted and eager). Correct with `MLX_DISABLE_COMPILE=1`, unchanged by `JAX_MPS_NO_OPTIMIZE=1`. Measured and reduced by the session that ran the #2274 parity sweep on an M4 Pro; the 30-line tengri-free reproducer is in the upstream issue.

tengri hits it in `components/agn/polar_dust.py:456`, `anisotropic_polar_luminosity`: `jnp.trapezoid(l_nu_absorbed[::-1], nu[::-1])` followed by `maximum(., 0)` returns exactly 0.0 on MPS, so the torus loses its far-IR graybody and its self-normalization redistributes the deficit: a torus-only model is flat x1.29 over 1-10 um and x0.067 at 100 um (x0.04 in the Herschel 250 band). Only the torus seams of the parity sweep failed, and every isolated-stage bisection agreed because none reached the polar-dust path.

A grep for the same shape in `src/tengri` finds three more `trapezoid` calls over reversed operands that are not yet measured on MPS and are candidates for the same failure:

- `components/agn/adaf.py:731` `jnp.trapezoid((total / _LSUN_ERG)[::-1], nu[::-1])`
- `components/agn/adaf.py:742` `jnp.trapezoid(total[::-1], nu[::-1])`
- `components/agn/unified.py:1265` `jnp.trapezoid(jnp.flip(l_disc_full), jnp.flip(nu))`

The other ~37 reversals in `src/tengri` (`[::-1]` on cumsums, SFH grids, `cosmology.py`, `dsps_wrapper.py`, `simulate.py`) do not combine the reversed array with a broadcast scalar on the same line, but `jnp.trapezoid` does internally (`* 0.5` on the pairwise sums), which is why the four `trapezoid` sites are the ones at risk.

## Options

1. Do nothing in tengri and document `MLX_DISABLE_COMPILE=1` as part of the Apple-GPU recipe (done in #2294; no speed cost, the workload is dispatch-bound). This is the current state.
2. Side-step the pattern at the four sites: `jnp.trapezoid` over the descending `nu` directly and take `jnp.abs` (or negate), which removes the reversal; float64 changes only by summation order (measure to 1e-12 before adopting). Worth doing only if MLX fixes the bug slowly and users run the default compile.
3. Add the four sites to `bench/scripts/benchmark_float32_mps_parity.py`'s per-seam reasons so a future MPS run names the site rather than the seam.

**Comment by suchethac** [comment]
Claiming the long-term fix (orchestrator session). Upstream jax-mps#232 is still open with zero comments, which meets this issue's own bar for adopting option 2 rather than waiting.

Plan:
- **Option 2** at all four sites: integrate over the descending `nu` directly and negate (not `abs` — negation keeps a wrong-ordering bug loud instead of silently rectifying it), removing the reversed-operand pattern that MLX compile mis-executes. Each site's `nu` ordering is verified at the call site before the rewrite, and float64 equivalence against the reversed form is measured to 1e-12 before adoption, per the issue text.
- **Option 3**: the four sites land in `bench/scripts/benchmark_float32_mps_parity.py`'s per-seam reasons so a future MPS run names the site, not the seam.
- Acceptance: CPU float64 equivalence test in-tree; MPS validation with default compile (torus-only model no longer flat x1.29 / x0.067) is being coordinated with the session that holds the M4 Pro reproducer environment, since CI has no MPS runner.

Option 1 (#2294's `MLX_DISABLE_COMPILE=1` recipe) stays documented as the recipe for released versions regardless.

**Comment by suchethac** [comment]
Fix is up: PR #2298 (`fix/2295-reversed-trapezoid`, head 12bf85956e45).

- Option 2 at all four sites, in the measured-safe form: negate the **scalar** trapezoid result over the descending `nu` — no `[::-1]`/`flip`/`rev` survives at any site (negating a reversed *array* is equally miscompiled, per the follow-up measurements posted to jax-mps#232).
- Float64 equivalence with the old form: ≤ 2.2e-16 max relative difference per site.
- Regression tests: grid-identity both orderings, reference-equivalence through `anisotropic_polar_luminosity` at rtol 1e-12, and a source scan forbidding the pattern in `src/tengri` (detector verified against the pre-fix tree: flags exactly these four sites).
- Option 3: `_REWRITTEN_SITES` ledger printed under the MPS parity benchmark's report.
- Option 1 unchanged: `MLX_DISABLE_COMPILE=1` stays in the Apple-GPU recipe — VJPs emit `lax.rev` regardless, so this fix corrects the forward paths only; gradient probes are being recorded as part of acceptance.

MPS acceptance (forward probes + recorded gradients + full 6-seam sweep, M4 Pro, default compile) is running against that SHA in the session holding the jax-mps environment. Merge is sequenced after #2291 by agreement with its session. This issue closes when the MPS numbers land and the PR merges.

**Comment by suchethac** [comment]
Fixed and closed. PR #2298 merged as a0e9ab3d78ed: all four reversed-operand trapezoids (polar_dust anisotropic luminosity, adaf float32+float64 normalizations, unified disc L_bol) now negate the scalar integral over the descending nu — the measured-safe form under default MLX compile. Acceptance on the M4 Pro (full tables on the PR): every forward probe ratio 1.0000 (torus far-IR restored from exactly 0.0, adaf from a 2732x collapse), the recorded polar_ebv gradient exact (trapezoid's VJP lowers to pad/add, not rev), float64 equivalence <= 2.2e-16 per site, and the parity sweep's two torus seams went from FAIL to their compile-off values. Option 3's per-site ledger prints under the sweep report; the source scan forbids the pattern tree-wide. Option 1 stands: MLX_DISABLE_COMPILE=1 stays in the recipe until upstream jax-mps#232 closes (VJPs elsewhere still emit lax.rev); the sweep's one remaining seam failure was bisected to a stale float64 reference, tracked as #2300.
